### PR TITLE
Templates

### PR DESCRIPTION
**Pull request by bcrochet**
Move the definition of some structs to just be templates instead.

### CODE REVIEW

**Comment by mgoerens** [comment]
Thanks @bcrochet for the PR ! A couple of remarks:

1) In this project we create currently up to 4 Deployments (Synapse, Heisendridge, MautrixSignal, Signald). There might be more in the future if / when we add more bridges and features. I would imagine it to be overly complicated to create a unique "deployment template" which can be configured for each of those use cases. Besides the image name and tag, there are other differences. For instance some use initContainers. Is converging to a unique template for all deployments a ideal goal in your opinion ? Vs having a "Synapse deployment template", a "Heisenbridge deployment template" and so on ? If the second, is this worth the effort ?

2) In a way related to (1), I like the face that you use `{{ .Values.Name }}` and not `{{ .Synapse.Name }}` in the templates as this convey a more generic usage.

3) in `ResourceFromTemplate`, could one go with `client.Object` instead of `any` ? Something like:
```go
func [T client.Object, R client.Object](t *T, name string, scheme *runtime.Scheme, fs fs.ReadFileFS) (*R, error)
```

4) Have you tested it since last week ? :) There seem to be an issue with `could not get template: could not read homeserver template: open homeserver.yaml: file does not exist` (haven't investigated, just seen this log from the [Github actions](https://github.com/opdev/synapse-operator/actions/runs/4660650282/jobs/8248964239?pr=91))

5) Wild though, but do you think using such templates in Operator project would simplify transforming a Helm chart into a Go Operator ? i.e. reusing the existing Helm template almost out of the box with minimal changes ? (assuming one would prefer a Go vs a Helm based Operator).

**Comment by mgoerens** [comment]
Answering my own questions, especially reg. 1)

For `Deployments` in particular, I think they differ too much from one another in order to have a unique "all purpose" deployment template. What we could have though, is a base template for general purposes, and further customize it after having loaded it. Something like:
```go

func (r *SynapseReconciler) deploymentForSynapse(s *synapsev1alpha1.Synapse, fs fs.ReadFileFS) (*appsv1.Deployment, error) {
	dep, err := templates.ResourceFromTemplate[synapsev1alpha1.Synapse, appsv1.Deployment](s, "deployment", r.Scheme, fs)
	if err != nil {
		return nil, fmt.Errorf("could not get template: %v", err)
	}

	dep.Spec.initContainers = ...
	dep.Spec ...
}
```

Maybe I'm overthinking this. We don't *need* a unique template, it's perfectly fine to have a "Synapse deployment template", a "Heisenbridge deployment template", and so on.

For other resources though, such as `PVCs` or `Services`, a unique template would be possible.

Besides, I definitely see the value is both the initial creation, and the later readability of using a template vs using go structs.

**Comment by bcrochet** [comment]
> Thanks @bcrochet for the PR ! A couple of remarks:
> 
>     1. In this project we create currently up to 4 Deployments (Synapse, Heisendridge, MautrixSignal, Signald). There might be more in the future if / when we add more bridges and features. I would imagine it to be overly complicated to create a unique "deployment template" which can be configured for each of those use cases. Besides the image name and tag, there are other differences. For instance some use initContainers. Is converging to a unique template for all deployments a ideal goal in your opinion ? Vs having a "Synapse deployment template", a "Heisenbridge deployment template" and so on ? If the second, is this worth the effort ?
> 

There are ways of either converging the templates or having them be separate. I could see separating them at first, and then discovering the efficiencies later.

>     2. In a way related to (1), I like the face that you use `{{ .Values.Name }}` and not `{{ .Synapse.Name }}` in the templates as this convey a more generic usage.
> 

This and #5 are related. I'll answer there.

>     3. in `ResourceFromTemplate`, could one go with `client.Object` instead of `any` ? Something like:
> 
> 
> ```go
> func [T client.Object, R client.Object](t *T, name string, scheme *runtime.Scheme, fs fs.ReadFileFS) (*R, error)
> ```
> 

I'd rather create a type for the constraint. Or, we can remove the generic altogether, and just have that be the type.


>     4. Have you tested it since last week ? :) There seem to be an issue with `could not get template: could not read homeserver template: open homeserver.yaml: file does not exist` (haven't investigated, just seen this log from the [Github actions](https://github.com/opdev/synapse-operator/actions/runs/4660650282/jobs/8248964239?pr=91))
>

I have not. :)
 
>     5. Wild though, but do you think using such templates in Operator project would simplify transforming a Helm chart into a Go Operator ? i.e. reusing the existing Helm template almost out of the box with minimal changes ? (assuming one would prefer a Go vs a Helm based Operator).

That was my thought with using .Values in the templates. I could see a sharing of templates like that.

**Comment by mgoerens** [comment]
After some further thinking I'd be in favor of implementing fully in this project. I'm willing to pick up where you left off and plan to first debug why the pipeline is currently failing, then implement templates for other child resources

**Comment by mgoerens** [comment]
After some further thinking I'd be in favor of implementing fully in this project. I'm going to pick up where you left off and plan to first debug why the pipeline is currently failing, then implement templates for other child resources

**Comment by mgoerens** [comment]
@bcrochet  I'd like to have your opinion on https://github.com/opdev/synapse-operator/pull/91/commits/75b10163190a23801362e1f4b2a952c32ed8fa54

Reasoning behind is as follow:
* We want to have reusable templates. For instance we want one template for creating Services for `Synapse`, `Heisenbridge`, and `MautrixSignal`.
* Some values for those child resources are different in all three cases (e.g. `TargetPort`), however this value is not present in the CR. Until now we are only accessing values from the CR (e.g. `Values.Name`)
* Hence the need to provide "extraValues". I'm not a huge fan though because we lose the nice coherent `Values` name.

Wdyt ?

**Comment by mgoerens** [comment]
> > @bcrochet I'd like to have your opinion on [75b1016](https://github.com/opdev/synapse-operator/commit/75b10163190a23801362e1f4b2a952c32ed8fa54)
> > Reasoning behind is as follow:
> > ```
> > * We want to have reusable templates. For instance we want one template for creating Services for `Synapse`, `Heisenbridge`, and `MautrixSignal`.
> > 
> > * Some values for those child resources are different in all three cases (e.g. `TargetPort`), however this value is not present in the CR. Until now we are only accessing values from the CR (e.g. `Values.Name`)
> > 
> > * Hence the need to provide "extraValues". I'm not a huge fan though because we lose the nice coherent `Values` name.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Wdyt ?
> 
> I'll address these things in a follow-on commit.

Great, looking forward to it !

Two other things to keep in mind:
* Not all child resources share the name and namespace of the parent CR. The Signald resources (a `Deployment` and a `PVC`) are getting appended with `-signald` to avoid conflict with the mautrix-signal resources.  The current implementation (using `{{ .Values.Name }}` and `{{ .Values.Namespace }}`) doesn't address this.
* The labels selector (in `Services` and `Deployments`) also need another handling, as the label key name also needs to be parametrized.

**Comment by mgoerens** [comment]
@bcrochet if you want to take a look, I believe I'm done here.

All what's left is to rebase on master to and fix merge conflicts. They are due to #94 . I'll look at it tomorrow.

File: internal/templates/templates.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	}
+	tmpl, err := template.New(name).Parse(string(resYaml))
+	if err != nil {
+		return nil, fmt.Errorf("could not parse %s template: %v", name, err)
+	}
+	var buf bytes.Buffer
+
+	err = tmpl.Execute(&buf, struct {
+		Values      *T
+		ExtraValues any
```
Comment (bcrochet) [inline internal/templates/templates.go:29]: Looks like I missed this?

File: controllers/synapse/mautrixsignal/signald_deployment.go [hunk 2]
```diff
@@ -66,44 +64,25 @@ func (r *MautrixSignalReconciler) reconcileSignaldDeployment(ctx context.Context
...
+	type deploymentExtraValues struct {
+		synapsev1alpha1.MautrixSignal
+		Labels map[string]string
+		Name   string
+	}
+
+	extraValues := deploymentExtraValues{
+		MautrixSignal: *ms,
+		Labels:        labelsForSignald(ms.Name),
+		Name:          GetSignaldResourceName(*ms),
```
Comment (bcrochet) [inline controllers/synapse/mautrixsignal/signald_deployment.go:77]: Why not just do this in the template itself?

File: internal/templates/signald_deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ .Values.Name }}
```
Comment (bcrochet) [inline internal/templates/signald_deployment.yaml:4]: I think the name template could just be here

```suggestion
  name: {{ .Values.Name }}-signald
```

File: controllers/synapse/mautrixsignal/signald_deployment.go (hunk 2 above)
Comment (mgoerens) [reply controllers/synapse/mautrixsignal/signald_deployment.go:77]: I'd argue it's best to pass those values for 2 reasons:
1. It paves the way for a potential future generic deployment templates.
2. It provides consistency for labels and name across the `Signald` resources.

File: internal/templates/templates.go (hunk 1 above)
Comment (mgoerens) [reply internal/templates/templates.go]: Right, I just removed it. Also the `r.Scheme` argument was unused in this function.

**Comment by mgoerens** [comment]
@bcrochet Rebased on master, squashed all commits, and fixed merge conflicts

File: controllers/synapse/mautrixsignal/signald_deployment.go (hunk 2 above)
Comment (bcrochet) [reply controllers/synapse/mautrixsignal/signald_deployment.go:77]: Fair enough. Does GetSignaldResourceName need to be exported? Is it used outside of this package? If so, should it live in a standalone package? (Not utils :D)

**Pull request merged**